File: javascript/packages/linter/src/rules/erb-no-inline-case-conditions.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ERBCaseNode, ERBCaseMatchNode, ERBWhenNode, ERBInNode, ParseResult, ParserOptions } from "@herb-tools/core"

class ERBNoInlineCaseConditionsVisitor extends BaseRuleVisitor {
  visitERBCaseNode(node: ERBCaseNode): void {
    this.checkConditions(node, "when")
    this.visitChildNodes(node)
  }

  visitERBCaseMatchNode(node: ERBCaseMatchNode): void {
    this.checkConditions(node, "in")
    this.visitChildNodes(node)
  }

  private checkConditions(node: ERBCaseNode | ERBCaseMatchNode, type: string): void {
    if (!node.conditions || node.conditions.length === 0) return

    for (const condition of node.conditions as (ERBWhenNode | ERBInNode)[]) {
      if (condition.tag_opening === null) {
        this.addOffense(
          `A \`case\` statement with \`${type}\` conditions in a single ERB tag cannot be reliably parsed, compiled, and formatted. Use separate ERB tags for \`case\` and its conditions (e.g., \`<% case x %>\` followed by \`<% ${type} y %>\`).`,
          node.location,
        )
        break
      }
    }
  }
}

export class ERBNoInlineCaseConditionsRule extends ParserRule {
  static ruleName = "erb-no-inline-case-conditions"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning",
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return { strict: false }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoInlineCaseConditionsVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_inline_case_conditions.rs
use herb::nodes::{AnyNode, ERBCaseMatchNode, ERBCaseNode};
use herb::Location;
use herb::Visitor;

rule_visitor!(ERBNoInlineCaseConditionsVisitor);
define_parser_rule!(
  ERBNoInlineCaseConditionsRule,
  "erb-no-inline-case-conditions",
  Warning,
  ERBNoInlineCaseConditionsVisitor,
  parser_options: { strict: false }
);

impl ERBNoInlineCaseConditionsVisitor {
  fn check_conditions(&mut self, conditions: &[AnyNode], location: &Location, condition_type: &str) {
    let has_inline_condition = conditions.iter().any(|condition| match condition {
      AnyNode::ERBWhenNode(when) => when.tag_opening.is_none(),
      AnyNode::ERBInNode(in_node) => in_node.tag_opening.is_none(),
      _ => false,
    });

    if has_inline_condition {
      self.add_offense(
        format!(
          "A `case` statement with `{}` conditions in a single ERB tag cannot be reliably parsed, compiled, and formatted. Use separate ERB tags for `case` and its conditions (e.g., `<% case x %>` followed by `<% {} y %>`).",
          condition_type, condition_type
        ),
        location.clone(),
      );
    }
  }
}

impl Visitor for ERBNoInlineCaseConditionsVisitor {
  fn visit_erb_case_node(&mut self, node: &ERBCaseNode) {
    self.check_conditions(&node.conditions, &node.location, "when");
    self.walk_erb_case_node(node);
  }

  fn visit_erb_case_match_node(&mut self, node: &ERBCaseMatchNode) {
    self.check_conditions(&node.conditions, &node.location, "in");
    self.walk_erb_case_match_node(node);
  }
}
